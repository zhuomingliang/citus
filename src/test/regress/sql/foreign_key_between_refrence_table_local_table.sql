CREATE SCHEMA fkey_reference_local_table;
SET search_path TO 'fkey_reference_local_table';

-- create test tables

CREATE TABLE local_table(l1 int);
CREATE TABLE reference_table(r1 int primary key);
SELECT create_reference_table('reference_table');

-- foreign key from local table to reference table --

-- this should fail
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1);

-- we do not support ALTER TABLE ADD CONSTRAINT fkey between local & reference table
-- within a transaction block, below block should fail
BEGIN;
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1);
COMMIT;

-- replicate reference table to coordinator
SELECT master_add_node('localhost', :master_port, groupId => 0);

-- we support ON DELETE CASCADE behaviour in "ALTER TABLE ADD fkey local_table (to reference_table) commands
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1) ON DELETE CASCADE;

-- show that on delete cascade works
INSERT INTO reference_table VALUES (11);
INSERT INTO local_table VALUES (11);
DELETE FROM reference_table WHERE r1=11;
SELECT * FROM local_table ORDER BY l1;

-- show that we support drop constraint
ALTER TABLE local_table DROP CONSTRAINT fkey_local_to_ref;

-- we support ON UPDATE CASCADE behaviour in "ALTER TABLE ADD fkey local_table (to reference table)" commands
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1) ON UPDATE CASCADE;

-- show that on delete cascade works
INSERT INTO reference_table VALUES (12);
INSERT INTO local_table VALUES (12);
UPDATE reference_table SET r1=13 WHERE r1=12;
SELECT * FROM local_table ORDER BY l1;

-- drop constraint for next commands
ALTER TABLE local_table DROP CONSTRAINT fkey_local_to_ref;

-- show that we are checking for foreign key constraint while defining

INSERT INTO local_table VALUES (2);

-- this should fail
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1);

INSERT INTO reference_table VALUES (2);

-- this should work
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1);

-- show that we are checking for foreign key constraint after defining

-- this should fail
INSERT INTO local_table VALUES (1);

INSERT INTO reference_table VALUES (1);

-- this should work
INSERT INTO local_table VALUES (1);

-- we do not support ALTER TABLE DROP CONSTRAINT fkey between local & reference table
-- within a transaction block, below block should fail
BEGIN;
ALTER TABLE local_table DROP CONSTRAINT fkey_local_to_ref;
COMMIT;

-- show that we do not allow removing coordinator when we have a fkey constraint
-- between a coordinator local table and a reference table
SELECT master_remove_node('localhost', :master_port);

-- drop and add constraint for next commands
ALTER TABLE local_table DROP CONSTRAINT fkey_local_to_ref;

ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1);

-- show that drop table without CASCADE does not error as we already append CASCADE
DROP TABLE reference_table;

-- drop local_table finally
DROP TABLE local_table;

-- TODO: drop them together after fixing the bug

-- create test tables

CREATE TABLE local_table(l1 int primary key);
CREATE TABLE reference_table(r1 int);
SELECT create_reference_table('reference_table');

-- remove master node from pg_dist_node
SELECT master_remove_node('localhost', :master_port);

-- foreign key from reference table to local table --

-- this should fail
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1);

-- we do not support ALTER TABLE ADD CONSTRAINT fkey between local & reference table
-- within a transaction block, below block should fail
BEGIN;
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1);
COMMIT;

-- replicate reference table to coordinator

SELECT master_add_node('localhost', :master_port, groupId => 0);

-- show that we are checking for foreign key constraint while defining

INSERT INTO reference_table VALUES (3);

-- this should fail
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1);

INSERT INTO local_table VALUES (3);

-- this should work
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1);

-- show that we are checking for foreign key constraint after defining

-- this should fail
INSERT INTO reference_table VALUES (4);

INSERT INTO local_table VALUES (4);

-- we do not support ON DELETE/UPDATE CASCADE behaviour in "ALTER TABLE ADD fkey reference_table (to local_table)" commands
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1) ON DELETE CASCADE;
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1) ON UPDATE CASCADE;

-- this should work
INSERT INTO reference_table VALUES (4);

-- we do not support ALTER TABLE DROP CONSTRAINT fkey between local & reference table
-- within a transaction block, below block should fail
BEGIN;
ALTER TABLE reference_table DROP CONSTRAINT fkey_ref_to_local;
COMMIt;

-- show that we do not allow removing coordinator when we have a fkey constraint
-- between a coordinator local table and a reference table
SELECT master_remove_node('localhost', :master_port);

-- show that we support drop constraint
ALTER TABLE reference_table DROP CONSTRAINT fkey_ref_to_local;

ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1);

-- show that drop table errors as expected
DROP TABLE local_table;

-- this should work
DROP TABLE local_table CASCADE;

-- drop reference_table finally
DROP TABLE reference_table;

-- TODO: drop them together after fixing the bug

-- TODO: test schema / table name escape
-- TODO: test create table behaviour, also implement error out conditions
-- TODO: partitioned tables

-- TEARDOWN -- -- -- --

-- print table reference table content to make sure we successfully inserted/deleted values
SELECT * FROM reference_table ORDER BY r1;

-- remove master node from pg_dist_node
SELECT master_remove_node('localhost', :master_port);

-- print table contents to make sure we successfully inserted/deleted values
SELECT * FROM local_table ORDER BY l1;
SELECT * FROM reference_table ORDER BY r1;

DROP SCHEMA fkey_reference_local_table;
