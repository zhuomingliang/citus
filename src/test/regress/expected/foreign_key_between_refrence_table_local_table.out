CREATE SCHEMA fkey_reference_local_table;
SET search_path TO 'fkey_reference_local_table';
-- create test tables
CREATE TABLE local_table(l1 int);
CREATE TABLE reference_table(r1 int primary key);
SELECT create_reference_table('reference_table');
 create_reference_table
---------------------------------------------------------------------

(1 row)

-- foreign key from local table to reference table --
-- this should fail
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1);
ERROR:  cannot ADD/DROP foreign key constraint
DETAIL:  Referenced table must be a distributed table or a reference table or a local table in coordinator.
HINT:  To ADD/DROP foreign constraint between reference tables and coordinator local tables, consider adding coordinator to pg_dist_node as well.
-- we do not support ALTER TABLE ADD CONSTRAINT fkey between local & reference table
-- within a transaction block, below block should fail
BEGIN;
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1);
ERROR:  cannot ADD/DROP foreign key constraint between coordinator local tables and reference tables in a transaction block, udf block, or distributed CTE subquery
COMMIT;
-- replicate reference table to coordinator
SELECT master_add_node('localhost', :master_port, groupId => 0);
NOTICE:  Replicating reference table "orders_reference" to the node localhost:xxxxx
NOTICE:  Replicating reference table "customer" to the node localhost:xxxxx
NOTICE:  Replicating reference table "nation" to the node localhost:xxxxx
NOTICE:  Replicating reference table "part" to the node localhost:xxxxx
NOTICE:  Replicating reference table "supplier" to the node localhost:xxxxx
NOTICE:  Replicating reference table "multi_outer_join_right_reference" to the node localhost:xxxxx
NOTICE:  Replicating reference table "multi_outer_join_third_reference" to the node localhost:xxxxx
NOTICE:  Replicating reference table "reference_table" to the node localhost:xxxxx
 master_add_node
---------------------------------------------------------------------
               4
(1 row)

-- we support ON DELETE CASCADE behaviour in "ALTER TABLE ADD fkey local_table (to reference_table) commands
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1) ON DELETE CASCADE;
-- show that on delete cascade works
INSERT INTO reference_table VALUES (11);
INSERT INTO local_table VALUES (11);
DELETE FROM reference_table WHERE r1=11;
SELECT * FROM local_table ORDER BY l1;
 l1
---------------------------------------------------------------------
(0 rows)

-- show that we support drop constraint
ALTER TABLE local_table DROP CONSTRAINT fkey_local_to_ref;
-- we support ON UPDATE CASCADE behaviour in "ALTER TABLE ADD fkey local_table (to reference table)" commands
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1) ON UPDATE CASCADE;
-- show that on delete cascade works
INSERT INTO reference_table VALUES (12);
INSERT INTO local_table VALUES (12);
UPDATE reference_table SET r1=13 WHERE r1=12;
SELECT * FROM local_table ORDER BY l1;
 l1
---------------------------------------------------------------------
 13
(1 row)

-- drop constraint for next commands
ALTER TABLE local_table DROP CONSTRAINT fkey_local_to_ref;
-- show that we are checking for foreign key constraint while defining
INSERT INTO local_table VALUES (2);
-- this should fail
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1);
ERROR:  insert or update on table "local_table" violates foreign key constraint "fkey_local_to_ref"
DETAIL:  Key (l1)=(2) is not present in table "reference_table_1190015".
INSERT INTO reference_table VALUES (2);
-- this should work
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1);
-- show that we are checking for foreign key constraint after defining
-- this should fail
INSERT INTO local_table VALUES (1);
ERROR:  insert or update on table "local_table" violates foreign key constraint "fkey_local_to_ref"
DETAIL:  Key (l1)=(1) is not present in table "reference_table_1190015".
INSERT INTO reference_table VALUES (1);
-- this should work
INSERT INTO local_table VALUES (1);
-- we do not support ALTER TABLE DROP CONSTRAINT fkey between local & reference table
-- within a transaction block, below block should fail
BEGIN;
ALTER TABLE local_table DROP CONSTRAINT fkey_local_to_ref;
ERROR:  cannot ADD/DROP foreign key constraint between coordinator local tables and reference tables in a transaction block, udf block, or distributed CTE subquery
COMMIT;
-- show that we do not allow removing coordinator when we have a fkey constraint
-- between a coordinator local table and a reference table
SELECT master_remove_node('localhost', :master_port);
ERROR:  cannot remove reference table placement from coordinator
DETAIL:  Local table "local_table" is involved in a foreign key constraint with refence table "reference_table", which has replica in coordinator
HINT:  DROP foreign key constraint between them or drop referenced table with DROP ... CASCADE.
-- drop and add constraint for next commands
ALTER TABLE local_table DROP CONSTRAINT fkey_local_to_ref;
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_ref FOREIGN KEY(l1) REFERENCES reference_table(r1);
-- show that drop table without CASCADE does not error as we already append CASCADE
DROP TABLE reference_table;
-- drop local_table finally
DROP TABLE local_table;
-- TODO: drop them together after fixing the bug
-- create test tables
CREATE TABLE local_table(l1 int primary key);
CREATE TABLE reference_table(r1 int);
SELECT create_reference_table('reference_table');
 create_reference_table
---------------------------------------------------------------------

(1 row)

-- remove master node from pg_dist_node
SELECT master_remove_node('localhost', :master_port);
 master_remove_node
---------------------------------------------------------------------

(1 row)

-- foreign key from reference table to local table --
-- this should fail
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1);
ERROR:  cannot ADD/DROP foreign key constraint
DETAIL:  Referenced table must be a distributed table or a reference table or a local table in coordinator.
HINT:  To ADD/DROP foreign constraint between reference tables and coordinator local tables, consider adding coordinator to pg_dist_node as well.
-- we do not support ALTER TABLE ADD CONSTRAINT fkey between local & reference table
-- within a transaction block, below block should fail
BEGIN;
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1);
ERROR:  cannot ADD/DROP foreign key constraint between coordinator local tables and reference tables in a transaction block, udf block, or distributed CTE subquery
COMMIT;
-- replicate reference table to coordinator
SELECT master_add_node('localhost', :master_port, groupId => 0);
NOTICE:  Replicating reference table "orders_reference" to the node localhost:xxxxx
NOTICE:  Replicating reference table "customer" to the node localhost:xxxxx
NOTICE:  Replicating reference table "nation" to the node localhost:xxxxx
NOTICE:  Replicating reference table "part" to the node localhost:xxxxx
NOTICE:  Replicating reference table "supplier" to the node localhost:xxxxx
NOTICE:  Replicating reference table "multi_outer_join_right_reference" to the node localhost:xxxxx
NOTICE:  Replicating reference table "multi_outer_join_third_reference" to the node localhost:xxxxx
NOTICE:  Replicating reference table "reference_table" to the node localhost:xxxxx
 master_add_node
---------------------------------------------------------------------
               5
(1 row)

-- show that we are checking for foreign key constraint while defining
INSERT INTO reference_table VALUES (3);
-- this should fail
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1);
ERROR:  insert or update on table "reference_table_1190016" violates foreign key constraint "fkey_ref_to_local"
DETAIL:  Key (r1)=(3) is not present in table "local_table".
INSERT INTO local_table VALUES (3);
-- this should work
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1);
-- show that we are checking for foreign key constraint after defining
-- this should fail
INSERT INTO reference_table VALUES (4);
ERROR:  insert or update on table "reference_table_1190016" violates foreign key constraint "fkey_ref_to_local"
DETAIL:  Key (r1)=(4) is not present in table "local_table".
INSERT INTO local_table VALUES (4);
-- we do not support ON DELETE/UPDATE CASCADE behaviour in "ALTER TABLE ADD fkey reference_table (to local_table)" commands
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1) ON DELETE CASCADE;
ERROR:  cannot create foreign key constraint
DETAIL:  Foreign key constraints from reference tables to coordinator local tables can only enforce RESTRICT or NO ACTION behaviour ON DELETE / UPDATE
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1) ON UPDATE CASCADE;
ERROR:  cannot create foreign key constraint
DETAIL:  Foreign key constraints from reference tables to coordinator local tables can only enforce RESTRICT or NO ACTION behaviour ON DELETE / UPDATE
-- this should work
INSERT INTO reference_table VALUES (4);
-- we do not support ALTER TABLE DROP CONSTRAINT fkey between local & reference table
-- within a transaction block, below block should fail
BEGIN;
ALTER TABLE reference_table DROP CONSTRAINT fkey_ref_to_local;
ERROR:  cannot ADD/DROP foreign key constraint between coordinator local tables and reference tables in a transaction block, udf block, or distributed CTE subquery
COMMIt;
-- show that we do not allow removing coordinator when we have a fkey constraint
-- between a coordinator local table and a reference table
SELECT master_remove_node('localhost', :master_port);
ERROR:  cannot remove reference table placement from coordinator
DETAIL:  Local table "local_table" is involved in a foreign key constraint with refence table "reference_table", which has replica in coordinator
HINT:  DROP foreign key constraint between them or drop referenced table with DROP ... CASCADE.
-- show that we support drop constraint
ALTER TABLE reference_table DROP CONSTRAINT fkey_ref_to_local;
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES local_table(l1);
-- show that drop table errors as expected
DROP TABLE local_table;
ERROR:  cannot drop table local_table because other objects depend on it
DETAIL:  constraint fkey_ref_to_local on table reference_table_1190016 depends on table local_table
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
-- this should work
DROP TABLE local_table CASCADE;
NOTICE:  drop cascades to constraint fkey_ref_to_local on table reference_table_1190016
-- drop reference_table finally
DROP TABLE reference_table;
-- TODO: drop them together after fixing the bug
-- TODO: test schema / table name escape
-- TODO: test create table behaviour, also implement error out conditions
-- TODO: partitioned tables
-- TEARDOWN -- -- -- --
-- print table reference table content to make sure we successfully inserted/deleted values
SELECT * FROM reference_table ORDER BY r1;
ERROR:  relation "reference_table" does not exist
-- remove master node from pg_dist_node
SELECT master_remove_node('localhost', :master_port);
 master_remove_node
---------------------------------------------------------------------

(1 row)

-- print table contents to make sure we successfully inserted/deleted values
SELECT * FROM local_table ORDER BY l1;
ERROR:  relation "local_table" does not exist
SELECT * FROM reference_table ORDER BY r1;
ERROR:  relation "reference_table" does not exist
DROP SCHEMA fkey_reference_local_table;
